### ISSUE TITLE
OneLL environment

### ISSUE THREAD

**Issue by ndangtt** [dimitri-rusin/oll_onemax#6]
There are a few things I'd like to mention regarding the current OneLL environment's implementation:

1. Observation space, action space, initial solution and total_number_of_function_evaluations should be created in the `reset` function. This is important when we move the multi-instance setting, as there's only a single training environment passed to the RL agent, and the switch to a new instance is typically done within `reset`. The constructor of OneLL can call that reset function at the end to initialise things (similar to our LeadingOnes benchmark: https://github.com/ndangtt/DACBench/blob/2eddd637e10e9ca7a01df0db00ba7efed369c801/dacbench/envs/theory.py#L396)

Another issue with the current implementation is that `self.total_number_of_function_evaluations` will be 2 if a OneLL is created and we call `reset` to start a new episode right after that.

2. The termination is based on two criteria: the current solution is optimal OR the cutoff time is reached. The current implementation of the `step` function misses the second criteria.

**Comment by dimitri-rusin** [comment]
> 1. This is important when we move the multi-instance setting,

Could it be that it's easier to concentrate on solving the single-instance setting really well first? So, only write code with the single-setting in mind? Or is this goal too easy to achieve? Once we have trained a really good policy for an instance of, say, 1,000 dimensions, that can compete with the theoretically derived policy, once we HAVE that, maybe having written the code just to handle the single-instance setting is already **a huge advantage** that we can leverage to possibly write code from scratch for the multiple-instance setting. Could it be?

**Comment by ndangtt** [comment]
I think the main contributions of this project is two-folds:

(1) Provide a new DAC benchmark based on OneLL on OneMax.

(2) Study the effectiveness of PPO on the benchmark. 

**Regarding (1):** 

We want to provide a DAC benchmark that facilitates future investigation of more advanced DAC techniques on the benchmark. The benchmark should be: (i) easy to use; and (ii) flexible. 

In optimisation settings, having a control policy that is generalisable across different instances is the standard (might not be the case with RL yet, but that's the ultimate aim of DAC). Therefore, I think it's important that we make sure the environment implemented takes that setting into account. 

The benchmark should also allow users to specify which instances they want to train the agent on, and in what order, and which instances are used for the evaluation.

The design of DACBench already takes those points into account. Implementing the benchmark in DACBench setting also helps to publicise the benchmark better. That was why I'd like to suggest that we implement the benchmark within DACBench. Of course if you want to keep the current implementation for simplicity, that's fine but at some point I think we need to move to the general setting, which makes the contribution more meaningful.

**Comment by ndangtt** [comment]
One question I have regarding the current implementation: what is 
`total_number_of_function_evaluations`? Is it #evaluations per episode? If that's true then it must be reset to zero in the `reset` function.

**Comment by dimitri-rusin** [comment]
> One question I have regarding the current implementation: what is `total_number_of_function_evaluations`? Is it #evaluations per episode? If that's true then it must be reset to zero in the `reset` function.

That's already removed. But now, for this next iteration I might introduce a similar variable. It tracks the number of function evaluations done across all episodes.

> The benchmark should also allow users to specify which instances they want to train the agent on, and in what order, and which instances are used for the evaluation.

So, let's clarify again one more thing: an **instance**. What is an instance? An instance could be:
* One episode, where a certain policy has been evaluated. We have a sequence of fitnesses and lambdas. This is one fixed sequence.
* A certain OneMax landscape. This is given by just defining the number n of dimensions. This can be evaluated in any way, we can use any policy to create an episode from this n-determined OneMax instance.
* An environment object. This is a Python object. Given this object, we can create several episodes. Each episode will have steps and in each step we will have several evaluations of the OneMax landscape. How many evaluations and at what point is determined by the random seed that is supplied to the environment object when it gets created.

So, basically the question is: what defines an instance? A random seed? A dimensionality? Percentage of how many bits are set to 1 in expectation in the initial point? Then, given all of these things the instance is just a combination of both a certain OneMax landscape (solely determined by dimensionality) PLUS an initialization of the OLL algorithm (determined by random seed as well as percentage of 1-bits used in the initial fitness). Is this what an instance is?

**Comment by ndangtt** [comment]
It's a great idea to consider #function evaluations as well, thanks Dimitri!

An instance represents the properties of the RL environment (the DAC benchmark) and can give us an indication about how difficult it is to learn in such environment. We can think about an instance as "parameters" of our OneLL-OneMax benchmark. The last one you mentioned matches this definition well. The parameters of an instance can include:
(1) OneMax dimension (n): the larger it is, the harder it is to learn
(2) The initial solution: the better this one is, the easier it is to learn
(3) The action space (the set of possible values for lambda): currently we set it as [1..sqrt(n)], but if we're to discretise lambda, the size of the portfolio of lambda values can be a factor that influence the difficulty of the instance. Let's ignore this for now though, as it would create some complication.
